widgets/src/Spinner.c

```diff
@@ -0,0 +1,330 @@
+}
+
+static void
+anaconda_spinner_disable_spin (AnacondaSpinner *spinner)
+{
+	if (spinner->priv->timeout_id)
+	{
+		g_source_remove (spinner->priv->timeout_id);
+		spinner->priv->timeout_id = 0;
+	}
+}
+
+static void
+anaconda_spinner_set_property (GObject *object,
+			guint property_id,
+			const GValue *value,
+			GParamSpec *pspec)
+{
+	switch (property_id) {
+		case PROP_ACTIVE:
+			anaconda_spinner_set_active (
+				ANACONDA_SPINNER (object),
+				g_value_get_boolean (value));
+			return;
+	}
+
+	G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
+}
+
+static void
+anaconda_spinner_get_property (GObject *object,
+			guint property_id,
+			GValue *value,
+			GParamSpec *pspec)
+{
+	switch (property_id) {
+		case PROP_ACTIVE:
+			g_value_set_boolean (
+				value,
+				anaconda_spinner_get_active (ANACONDA_SPINNER (object)));
+			return;
+	}
+
+	G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
+}
+
+static void
+anaconda_spinner_constructed (GObject *object)
+{
+	AnacondaSpinner *spinner;
+	GdkPixbuf *main_pixbuf;
+	gint xx, yy, width, height;
+	GError *error = NULL;
+        gchar *filename;
+
+	/* Chain up to parent's method. */
+	G_OBJECT_CLASS (anaconda_spinner_parent_class)->constructed (object);
+
+	spinner = ANACONDA_SPINNER (object);
+
+        filename = g_strdup_printf("%s/pixmaps/%s", anaconda_get_widgets_datadir(), MAIN_IMAGE_FILENAME);
+        main_pixbuf = gdk_pixbuf_new_from_file (filename, &error);
+        g_free(filename);
+
+	if (!main_pixbuf) {
+		g_warning ("%s: Failed to load image: %s", error ? error->message : "Unknown error", G_STRFUNC);
+		g_clear_error (&error);
+		return;
+	}
+
+	width = gdk_pixbuf_get_width (main_pixbuf);
+	height = gdk_pixbuf_get_height (main_pixbuf);
+
+	for (yy = 0; yy < height; yy += FRAME_SIZE) {
+		for (xx = 0; xx < width; xx+= FRAME_SIZE) {
+			GdkPixbuf *frame;
+
+			frame = gdk_pixbuf_new_subpixbuf (main_pixbuf, xx, yy, FRAME_SIZE, FRAME_SIZE);
+			if (frame)
+				spinner->priv->pixbufs = g_slist_prepend (spinner->priv->pixbufs, frame);
+		}
+	}
+
+	g_object_unref (main_pixbuf);
+
+	spinner->priv->pixbufs = g_slist_reverse (spinner->priv->pixbufs);
+
+	spinner->priv->current_frame = spinner->priv->pixbufs;
+	if (spinner->priv->pixbufs)
+		gtk_image_set_from_pixbuf (GTK_IMAGE (spinner), spinner->priv->pixbufs->data);
+}
+
+static void
+anaconda_spinner_dispose (GObject *object)
+{
+	/* This resets the timeout_id too */
+	anaconda_spinner_set_active (ANACONDA_SPINNER (object), FALSE);
+
+	/* Chain up to parent's method. */
+	G_OBJECT_CLASS (anaconda_spinner_parent_class)->dispose (object);
+}
+
+static void
+anaconda_spinner_finalize (GObject *object)
+{
+	AnacondaSpinner *spinner = ANACONDA_SPINNER (object);
+
+	g_slist_free_full (spinner->priv->pixbufs, g_object_unref);
+	spinner->priv->pixbufs = NULL;
+	spinner->priv->current_frame = NULL;
```

Comment:
If I remember that correctly, ``finalize`` is class' finalization and ``dispose`` is instance's finalization so the three lines above should go to ``dispose`` as these resources are per-instance.